libtock/services/alarm.c

```diff
@@ -145,27 +180,118 @@ static libtock_alarm_ticks_t* root_peek(void) {
   return root;
 }
 
+/** \brief Upcall for internal virtual alarms
+ *
+ * This upcall checks the ordered list of outstanding alarms for
+ * expired alarms, removes them from the list, and invokes their
+ * callbacks.
+ *
+ * Invariants:
+ * 1. The list of outstanding alarms is ordered by absolute expiration time.
+ * 2. The alarm event that invoked this upcall is for the current `head` of the list.
+ * 3. No alarms are added (or re-added) to the list between while
+ *    iterating through alarms
+ * 4. For each alarm, `alarm->dt < MAX_TICKS + 1`
+ *
+ * Corrollaries:
+ * - If the head of the list hasn't expired, no alarms in the tail of
+ *   the list have expired (1)
+ * - `scheduled` happens after `head->reference` (2)
+ * - `scheduled` happens before `head->reference + MAX_TICKS + 1` (4)
+ * - For each alarm in the list, `now` happens after
+ *   `alarm->reference` (2, 3)
+ * - For each alarm, `scheduled` cannot have cannot have wrapped
+ *   `alarm->reference` (1, 4)
+ * - For each alarm, if `scheduled` and `now` are not on the same side of
+ *   `alarm->reference`, the alarm must have expired (1, 4)
+ *
+ * Critically, this upcall cannot allow any alarms to be added to the
+ * ordered list while iterating by invoking upcalls, as that could
+ * violate invariant (3) and result in `now` happening before some
+ * `alarm->reference`.
+ *
+ * Some alarms that expire between `now` and the end of the upcall may
+ * be "missed", which may mean they are delivered later. They should
+ * still show up first in the list at the end, so will fire next.
+ */
 static void alarm_upcall(__attribute__ ((unused)) int   kernel_now,
-                         __attribute__ ((unused)) int   scheduled,
+                         int                            scheduled0,
                          __attribute__ ((unused)) int   unused2,
                          __attribute__ ((unused)) void* opaque) {
+  // `tocall` is a temporary list to keep track of expired alarms to call later.
+  libtock_alarm_ticks_t* tocall      = NULL;
+  libtock_alarm_ticks_t* tocall_last = NULL;
+
+  // We want `scheduled` as unsigned so wrapping math works out correctly.
+  uint32_t scheduled = scheduled0;
+
+  // Take the current tick value. We could use `kernel_now`, but would
+  // potentially unnecessarily delay some alarms.
+  uint32_t now;
+  libtock_alarm_command_read(&now);
+
+  // We know this upcall is associated with the head, so it's easier
+  // to deal with.
+  libtock_alarm_ticks_t* head = root_pop();
+
+  // Formally, we should be able to just add head to
+  // `tocall`, but let's just be defensive just in case there is an
+  // errant alarm.
+  assert(head != NULL && !is_within(scheduled, head->reference, head->dt));
+  head->next  = NULL;
+  head->prev  = tocall_last;
```

Comment:
Will this not always be `NULL`?

Reply:
`tocall_last`? It will, yeah. I'm fine changing this to `NULL`, though stylistically I could go both ways. i.e. this matches a pattern that we always set `->prev` to whatever the previously last thing in the list was, as opposed to special-casing the head. On the other hand, maybe it's unnecessary cognitive burden to read through this way.

Reply:
Fixed